### PR TITLE
check for matching alias before subproject slug

### LINKED ISSUE #2731: subproject alias 302 redirect to top-level project sharing alias' name
## Details
I have set up a subproject,
Subproject: cloudify-openstack-plugin-fh
Alias: openstack

* Project URL: http://cfy-rtd-demo.readthedocs.io/
* Build URL (if applicable): https://readthedocs.org/projects/cfy-rtd-demo/builds/5162299/
* Read the Docs username (if applicable): funkyhat

## Expected Result
When navigating to http://cfy-rtd-demo.readthedocs.io/projects/openstack

I expect to see (presumably via a redirect to http://cfy-rtd-demo.readthedocs.io/projects/openstack/en/sphinxify-rtd-demo/) my subproject's docs (`sphinxify-rtd-demo` is the current "active branch" for the subproject).

## Actual Result
I am redirected to http://openstack.readthedocs.io/en/latest/ which is unrelated to my project:
```
< HTTP/1.1 302 Found
* Server nginx/1.10.0 (Ubuntu) is not blacklisted
< Server: nginx/1.10.0 (Ubuntu)
< Date: Fri, 17 Mar 2017 13:00:04 GMT
< Content-Type: text/html; charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
< Vary: Accept-Language, Cookie
< Location: http://openstack.readthedocs.io/en/latest/
< Content-Language: en
< X-Fallback: True
< X-Served: Django
< X-Deity: web03
```

I have tried rebuilding the main project, which produced no change.

This seems potentially related to #1602

### PR DESCRIPTION
**Pull request by funkyHat**
fixes #2731

### CODE REVIEW

File: readthedocs/rtd_tests/tests/test_subprojects.py [hunk 1]
```diff
@@ -77,16 +77,27 @@ def setUp(self):
...
+    def test_resolver_subproject_alias(self):
         with override_settings(USE_SUBDOMAIN=False):
             resp = self.client.get('/docs/pip/projects/sub_alias/')
             self.assertEqual(resp.status_code, 302)
             self.assertEqual(
                 resp._headers['location'][1],
                 'http://readthedocs.org/docs/pip/projects/sub_alias/ja/latest/'
             )
+
+    def test_resolver_subproject_subdomain_alias(self):
```
Comment (funkyHat) [inline readthedocs/rtd_tests/tests/test_subprojects.py]: This test causes `test_resolver_subproject_alias` to fail if it runs first ("fixed" by creatively renaming so it runs after).
I had a bit of a look around to see if there's a cache I can easily clean but couldn't find anything.
Depends how much of an issue you think this is, but if you're worried & can give me a pointer I'll happily update it to dtrt

File: readthedocs/rtd_tests/tests/test_subprojects.py [hunk 2]
```diff
@@ -152,24 +152,34 @@ def setUp(self):
...
+        with override_settings(USE_SUBDOMAIN=True):
```
Comment (safwanrahman) [inline readthedocs/rtd_tests/tests/test_subprojects.py]: Can you use `override_settings` as decorator?

Comment (funkyHat) [reply readthedocs/rtd_tests/tests/test_subprojects.py]: done :)

**Comment by funkyHat** [comment]
Do you need anything else before this can be merged? Happy to make any required changes

File: readthedocs/rtd_tests/tests/test_subprojects.py [hunk 3]
```diff
@@ -152,24 +152,36 @@ def setUp(self):
             self.owner = create_user(username='owner', password='test')
             self.tester = create_user(username='tester', password='test')
             self.pip = fixture.get(Project, slug='pip', users=[self.owner], main_language_project=None)
-            self.subproject = fixture.get(Project, slug='sub', language='ja',
-                                          users=[ self.owner],
-                                          main_language_project=None)
-            self.translation = fixture.get(Project, slug='trans', language='ja',
-                                           users=[ self.owner],
-                                           main_language_project=None)
+            self.subproject = fixture.get(Project, slug='sub', language='ja', users=[
```
Comment (berkerpeksag) [inline readthedocs/rtd_tests/tests/test_subprojects.py]: I'd revert these style changes if the linting tool is happy with the old style.

File: readthedocs/rtd_tests/tests/test_resolver.py [hunk 4]
```diff
@@ -385,3 +385,32 @@ def test_resolver_public_domain_overrides(self):
             self.assertEqual(url, 'http://docs.foobar.com/en/latest/')
             url = resolve(project=self.pip, private=False)
             self.assertEqual(url, 'http://docs.foobar.com/en/latest/')
+
+
+class ResolverAltSetUp(object):
+
+    def setUp(self):
```
[the remark this answers is not in the document]
Comment (funkyHat) [reply readthedocs/rtd_tests/tests/test_resolver.py]: It does save 8s (out of ~20 for the module 👍), but it also seems to change the results:
```________________________________ SmartResolverPathTests.test_resolver_subproject_subdomain _________________________________

self = <readthedocs.rtd_tests.tests.test_resolver.SmartResolverPathTests testMethod=test_resolver_subproject_subdomain>

    def test_resolver_subproject_subdomain(self):
        with override_settings(USE_SUBDOMAIN=False):
            import pdb; pdb.set_trace()
            url = resolve_path(project=self.subproject, filename='index.html')
>           self.assertEqual(url, '/docs/pip/projects/sub/ja/latest/')
E           AssertionError: u'/docs/pip/projects/sub/' != '/docs/pip/projects/sub/ja/latest/'

rtd_tests/tests/test_resolver.py:105: AssertionError
_______________________________ SmartResolverPathTestsAlt.test_resolver_subproject_subdomain _______________________________

self = <readthedocs.rtd_tests.tests.test_resolver.SmartResolverPathTestsAlt testMethod=test_resolver_subproject_subdomain>

    def test_resolver_subproject_subdomain(self):
        with override_settings(USE_SUBDOMAIN=False):
            import pdb; pdb.set_trace()
>           url = resolve_path(project=self.subproject, filename='index.html')
E           AssertionError: u'/docs/pip/projects/sub/' != '/docs/pip/projects/sub/ja/latest/'```
I'm not really sure what's going on to cause that, unfortunately. Any clues?

[the remark this answers is not in the document]
Comment (berkerpeksag) [reply readthedocs/rtd_tests/tests/test_resolver.py]: Is `SmartResolverPathTestsAlt.test_resolver_subproject_subdomain` the only failing test? `SmartResolverPathTests` is the base class of `SmartResolverPathTestsAlt` and `ResolverBase` is the base class of `SmartResolverPathTests`. There is a `setUp` method defined in `ResolverBase`:

https://github.com/rtfd/readthedocs.org/blob/6a669349cf5349087a2369de8a6011b434ac858e/readthedocs/rtd_tests/tests/test_resolver.py#L18-L26

That method can be cause of the test failure.

Adding

```py
setUp = None
```

to `SmartResolverPathTestsAlt` may help.

There is also the following warning in `setUpTestData` documentation:

> Be careful not to modify any objects created in setUpTestData() in your test methods. Modifications to in-memory objects from setup work done at the class level will persist between test methods. If you do need to modify them, you could reload them in the setUp() method with refresh_from_db(), for example.

And we indeed do modify objects created in `setUpTestData`. For example:

https://github.com/rtfd/readthedocs.org/blob/6a669349cf5349087a2369de8a6011b434ac858e/readthedocs/rtd_tests/tests/test_resolver.py#L61

[the remark this answers is not in the document]
Comment (funkyHat) [reply readthedocs/rtd_tests/tests/test_resolver.py]: `test_resolver_subproject_subdomain ` from both `SmartResolverPathTests` and `SmartResolverPathTestsAlt` (i.e. yes, the one defined test).
These 2 failures are happening with both `setUp` methods (i.e. in `ResolverBase` and `ResolverAltSetUp`) converted to `seUpTestData` classmethods.

Are you happy to leave this optimisation for now? Reworking all the modifications seems like overkill for 8s savings (maybe should be a separate PR in any case)?

[the remark this answers is not in the document]
Comment (berkerpeksag) [reply readthedocs/rtd_tests/tests/test_resolver.py]: Yes, let's leave it to a separate PR.

File: readthedocs/core/views/serve.py [hunk 5]
```diff
@@ -67,10 +67,7 @@ def inner_view(request, subproject=None, subproject_slug=None, *args, **kwargs):
                 )
                 subproject = rel.child
             except (ProjectRelationship.DoesNotExist, KeyError):
-                try:
-                    subproject = Project.objects.get(slug=subproject_slug)
-                except Project.DoesNotExist:
-                    raise Http404
+                get_object_or_404(Project, slug=subproject_slug)
```
Comment (berkerpeksag) [inline readthedocs/core/views/serve.py]: Shouldn't this line be read as

```py
subproject = get_object_or_404(Project, slug=subproject_slug)
```

?

**Comment by tannewt** [comment]
I just ran into this as well. I would love to see this fixed so we don't need to encode language and version into URLs unnecessarily.

**Comment by ericholscher** [comment]
Can you include a link to something that reproduces this issue currently? There are likely other bugs that are causing this code to be hit.

**Comment by ericholscher** [comment]
Ah, I found https://github.com/rtfd/readthedocs.org/issues/2731 -- will give it a peek.

**Comment by ericholscher** [comment]
Looks good. Thanks for the patch -- sorry it took so long to review!

**Pull request merged**